packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,216 @@
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountId: string,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    const resolvedConditionType = this.resolveConditionType_(data)
+
+    if (!resolvedConditionType) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Missing one of products, collections, tags, types or customer groups in data`
+      )
+    }
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+        const discountRepo: DiscountRepository = manager.getCustomRepository(
+          this.discountRepository_
+        )
+
+        if (data.id) {
+          // throw if condition does not exist
+          const condition = await this.retrieve(data.id)
+
+          return await discountConditionRepo.addConditionResources(
+            condition.id,
+            resolvedConditionType.resource_ids,
+            resolvedConditionType.type,
+            true
+          )
+        }
+
+        const discount = await discountRepo.findOne({
+          where: { id: discountId },
+          relations: ["rule", "rule.conditions"],
+        })
+
+        if (!discount) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Discount with ${discountId} was not found`
+          )
+        }
+
+        const created = discountConditionRepo.create({
+          discount_rule_id: discount.rule_id,
+          operator: data.operator,
+          type: resolvedConditionType.type,
+        })
+
+        const discountCondition = await discountConditionRepo.save(created)
+
+        return await discountConditionRepo.addConditionResources(
+          discountCondition.id,
+          resolvedConditionType.resource_ids,
+          resolvedConditionType.type
+        )
+      },
+      async (err: { code: string }) => {
+        if (err.code === PostgresError.DUPLICATE_ERROR) {
+          // A unique key constraint failed meaning the combination of
+          // discount rule id, type, and operator already exists in the db.
+          throw new MedusaError(
+            MedusaError.Types.DUPLICATE_ERROR,
+            `Discount Condition with operator '${data.operator}' and type '${resolvedConditionType?.type}' already exist on a Discount Rule`
+          )
+        }
+      }
+    )
+  }
+
+  async remove(discountConditionId): Promise<void> {
```

Comment:
the return type from the method implementation should be `Promise<DiscountCondition | void>`